pkg/doc/verifiable/credential.go

```diff
@@ -340,32 +405,81 @@ func WithUnsecuredJWTDecoding() CredentialOpt {
 	}
 }
 
-func decodeIssuer(data []byte, credential *Credential) error {
-	issuerID, issuerName, err := issuerFromBytes(data)
-	if err != nil {
-		return fmt.Errorf("JSON unmarshalling of Verifiable Credential Issuer failed: %w", err)
+func decodeIssuer(rc *rawCredential) (Issuer, error) {
+	getStringEntry := func(m map[string]interface{}, k string) (string, error) {
+		v, exists := m[k]
+		if !exists {
+			return "", nil
+		}
+
+		s, valid := v.(string)
+		if !valid {
+			return "", fmt.Errorf("value of key '%s' is not a string", k)
+		}
+
+		return s, nil
 	}
 
-	credential.Issuer = Issuer{ID: issuerID, Name: issuerName}
-	return nil
-}
+	switch iss := rc.Issuer.(type) {
+	case string:
+		return Issuer{ID: iss}, nil
+	case map[string]interface{}:
+		id, err := getStringEntry(iss, "id")
+		if err != nil {
+			return Issuer{}, err
+		}
+		if id == "" {
+			return Issuer{}, errors.New("issuer ID is not defined")
+		}
 
-func decodeType(data []byte, vc *Credential) error {
-	single := typeSingle{}
-	err := json.Unmarshal(data, &single)
-	if err == nil {
-		vc.Type = single.Type
-		return nil
+		name, err := getStringEntry(iss, "name")
+		if err != nil {
+			return Issuer{}, err
+		}
+
+		return Issuer{
+			ID:   id,
+			Name: name,
+		}, nil
+	default:
+		return Issuer{}, errors.New("unsupported format of issuer")
 	}
+}
 
-	multiple := typeMultiple{}
-	err = json.Unmarshal(data, &multiple)
-	if err == nil {
-		vc.Type = multiple.Types
-		return nil
+func decodeType(rc *rawCredential) ([]string, error) {
+	switch rType := rc.Type.(type) {
+	case string:
+		return []string{rType}, nil
+	case []interface{}:
+		types, err := stringSlice(rType)
+		if err != nil {
+			return nil, fmt.Errorf("failed to map credential types: %w", err)
+		}
+		return types, nil
+	default:
+		return nil, errors.New("credential type of unknown type")
 	}
+}
```

Comment:
I do not like this approach with different types `string and []interface{}`, are we able to change this logic? We can use comma-separated values instead of different types e.g `type1` or  `type1,type3,...` or only slice [type1] or  [type1,type2]
@troyronda @rolsonquadras @fqutishat

Reply:
hey @soluchok 
According to [Verifiable Credentials spec](https://w3c.github.io/vc-data-model/#types), the `type` can be defined as either a single string value or array. We keep it raw in `rawCredential` struct (into which VC JSON is unmarshalled) and then decode it to universal and convenient `[]string` to be used in `Credential` struct.

Reply:
I would, at least, add some comments in the code. It isn’t entirely clear to a reader when they see these case statements.

Reply:
Sure.

Reply:
Done.